### DISCUSSION TITLE
move group c++ interface with ros2 action server

### DISCUSSION THREAD

**Discussion by archat** [moveit/moveit2#1534] [Q&A]
Hi All,

I started learning ROS2 and moveit2 recently, so my question might be very basic or some solution might exist already which i am still searching for. 

I am using panda robot and move group c++ interface as is given from the below link
https://moveit.picknik.ai/foxy/doc/move_group_interface/move_group_interface_tutorial.html
Also tried the ROS2 action server example as given below
https://docs.ros.org/en/foxy/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.html#compiling-the-action-server

Now i am planning to integrate both the code as single code, i need an action server to be running and i can request my goal from action client like close/open the gripper, move to home, move to point1, or ask to perfrom certain task like pick object1, pick object2 etc. 

When i tried to put it together i stuck with the node creation since action server code looks entirely different then normal c++ usage without main function. And for move group i need a node created and saved in a variable move_group_node so that i can pass this for  planning interface. I am kind of stuck between adapting both together, this is more likely on the c++ usage and integration but still i would like to ask this if any one has created such functionality before so that i can get some help on this quickly. So Request to provide me the option on how i can get it together.

Thanks in Advance.

**Comment by sjahr** [comment]
As stated in the tutorial "This interface communicates over ROS topics, services, and actions to the [MoveGroup](https://github.com/ros-planning/moveit2/blob/main/moveit_ros/move_group/src/move_group.cpp) node.", the move_group_interface already implements the actions clients for you and abstracts with interfaces like `move_group.plan(my_plan)`. Of course you can also directly communicate to the move_group with your own action server. You find the action servers offered by the move_group by running `ros2 action list` on your terminal when the tutorial is launched. You could also take a look into the source code of the [MoveGroupInterface](https://github.com/ros-planning/moveit2/tree/main/moveit_ros/planning_interface/move_group_interface) in the moveit2. If you struggle with the c++ usage, I'd recommend to read the tutorial code and/or try to implement easier examples based on the ros2 tutorials yourself :) If you have general ROS2  + C++ questions, [ROSAnswers](https://answers.ros.org/questions/) is the better place to ask those.
